x/pubkey/keeper/endblock.go

```diff
@@ -0,0 +1,155 @@
+	"github.com/sedaprotocol/seda-chain/x/pubkey/types"
+)
+
+func (k Keeper) EndBlock(ctx sdk.Context) (err error) {
+	// Use defer to prevent returning an error, which would cause
+	// the chain to halt.
+	defer func() {
+		// Handle a panic.
+		if r := recover(); r != nil {
+			k.Logger(ctx).Error("recovered from panic in pubkey end block", "err", r)
+		}
+		// Handle an error.
+		if err != nil {
+			k.Logger(ctx).Error("error in pubkey end block", "err", err)
+		}
+		err = nil
+	}()
+
+	scheme, err := k.GetProvingScheme(ctx, utils.SEDAKeyIndexSecp256k1)
+	if err != nil {
+		return err
+	}
+
+	// If the proving scheme is already activated, do nothing.
+	if scheme.IsActivated {
+		return
+	}
+
+	// Process activation in progress.
+	activationInProgress := scheme.ActivationHeight != types.DefaultActivationHeight
+	if activationInProgress && ctx.BlockHeight() >= scheme.ActivationHeight {
+		err = k.JailValidators(ctx, utils.SEDAKeyIndexSecp256k1)
+		if err != nil {
+			return err
+		}
+
+		scheme.IsActivated = true
+		scheme.ActivationHeight = types.DefaultActivationHeight
+		err = k.SetProvingScheme(ctx, scheme)
+		if err != nil {
+			return err
+		}
+
+		k.Logger(ctx).Info("proving scheme activated", "key_index", utils.SEDAKeyIndexSecp256k1)
+		return
+	}
+
+	// Check the public key registration rate and start the activation
+	// process if the rate has reached the threshold. If the activation
+	// process is already in progress and the threshold is not met,
+	// cancel the activation process.
+	met, err := k.CheckKeyRegistrationRate(ctx, utils.SEDAKeyIndexSecp256k1)
+	if err != nil {
+		return err
+	}
+	if !activationInProgress && met {
+		err = k.StartProvingSchemeActivation(ctx, utils.SEDAKeyIndexSecp256k1)
+		if err != nil {
+			return err
+		}
+	} else if activationInProgress && !met {
+		err = k.CancelProvingSchemeActivation(ctx, utils.SEDAKeyIndexSecp256k1)
+		if err != nil {
+			return err
+		}
+	}
+	return
+}
+
+// CheckKeyRegistrationRate checks if the current registration rate of
+// public keys of the given key scheme exceeds the threshold.
+func (k Keeper) CheckKeyRegistrationRate(ctx sdk.Context, keyIndex utils.SEDAKeyIndex) (bool, error) {
+	// If the sum of the voting power has reached 80%, enable secp256k1
+	// proving scheme.
+	totalPower, err := k.stakingKeeper.GetLastTotalPower(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	var powerSum uint64
+	err = k.stakingKeeper.IterateLastValidatorPowers(ctx, func(valAddr sdk.ValAddress, power int64) (stop bool) {
+		_, err := k.GetValidatorKeyAtIndex(ctx, valAddr, keyIndex)
+		if err != nil {
+			if errors.Is(err, collections.ErrNotFound) {
+				return false
+			}
+			panic(err)
+		}
+		//nolint:gosec // G115: We shouldn't get negative power anyway.
+		powerSum += uint64(power)
+		return false
+	})
+	if err != nil {
+		return false, err
+	}
+
+	//nolint:gosec // G115: We shouldn't get negative power anyway.
+	requiredPower := uint64(totalPower.Int64()*100*4/5 + 1)
```

Comment:
Should we have this 80% be a governance parameter?
The only constraint I see it needs to have is that it needs to be bigger than 2/3, right?

Reply:
Or at least a constant somewhere 😅